File: ios/client/Models/Metrics.swift
//
//  Metrics.swift
//  client
//
//  Created by Nadir Muzaffar on 5/11/20.
//  Copyright © 2020 Nadir Muzaffar. All rights reserved.
//

import Foundation

struct MuscleStat: Codable {
    let muscle: String
    let reps: Int
}

struct MetricsTopLevel: Codable {
    let distance: Float
    let sets: Int
    let reps: Int
    let secondsElapsed: Int
    let calories: Int
    
    enum CodingKeys: String, CodingKey {
        case secondsElapsed = "seconds_elapsed"
        case sets, reps, distance, calories
    }
}

struct MetricMuscle: Codable {
    let name: String
    let usage: String
    let reps: Int
}

struct Metric: Codable {
    let muscles: [MetricMuscle]
    let topLevel: MetricsTopLevel
    
    enum CodingKeys: String, CodingKey {
        case topLevel = "top_level"
        case muscles
    }
}

File: server/models/metrics.go
package models

const (
	TargetMuscle               = "TargetMuscle"
	SynergistMuscle            = "SynergistMuscle"
	StabilizerMuscle           = "StabilizerMuscle"
	DynamicStabilizerMuscle    = "DynamicStabilizerMuscle"
	AntagonistStabilizerMuscle = "AntagonistStabilizerMuscle"
	DynamicArticulationMuscle  = "DynamicArticulation"
	StaticArticulationMuscle   = "StaticArticulation"
)

// TODO: should metric be associated with workoutID or a time range?
type Metric struct {
	Model
	Muscles   []MetricMuscle `json:"muscles"`
	TopLevel  MetricTopLevel `json:"top_level"`
	WorkoutID uint           `json:"-" gorm:"type:int REFERENCES workouts(id) ON DELETE CASCADE"`
}

func (Metric) TableName() string {
	return "metrics"
}

type MetricTopLevel struct {
	HiddenModel
	Distance       float32 `json:"distance"`
	Sets           int     `json:"sets"`
	Reps           int     `json:"reps"`
	SecondsElapsed uint    `json:"seconds_elapsed"`
	Calories       int     `json:"calories"`
	MetricID       uint    `json:"metric_id" gorm:"type:int REFERENCES metrics(id) ON DELETE CASCADE"`
}

func (MetricTopLevel) TableName() string {
	return "metrics_top_level"
}

type MetricMuscle struct {
	HiddenModel
	Name     string `json:"name"`
	Usage    string `json:"usage"`
	Reps     int    `json:"reps"`
	MetricID uint   `json:"metric_id" gorm:"type:int REFERENCES metrics(id) ON DELETE CASCADE"`
}

func (MetricMuscle) TableName() string {
	return "metrics_muscle"
}
